File: enso-cell-count/frontend/lib/macrodissection/candidates.ts
// Client-side candidate ROI auto-suggester.
//
// Mirrors ``enso_purity.macrodissection.candidates`` so the static demo
// (no backend) can still surface candidate macrodissection regions.

import type { Point2D, TileArrays, ThresholdProfile } from "./types";

export interface ClientCandidate {
  rank: number;
  score: number;
  bbox_thumb_px: readonly [number, number, number, number];
  polygon: readonly Point2D[];
  purity_point: number;
  total_nuclei_point: number;
  tumor_nuclei_point: number;
  adequacy_probability: number;
}

function erfApprox(x: number): number {
  // Abramowitz & Stegun 7.1.26 — max error 1.5e-7.
  const sign = x < 0 ? -1 : 1;
  const ax = Math.abs(x);
  const t = 1 / (1 + 0.3275911 * ax);
  const y =
    1 -
    (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t - 0.284496736) *
      t +
      0.254829592) *
      t *
      Math.exp(-ax * ax);
  return sign * y;
}

function proxyAdequacyProb(
  purityPoint: number,
  puritySd: number,
  tumorPoint: number,
  tumorSd: number,
  purityMin: number,
  tumorCellsMin: number,
): number {
  const zPurity =
    puritySd <= 0
      ? purityPoint >= purityMin
        ? 8
        : -8
      : (purityPoint - purityMin) / puritySd;
  const zCells =
    tumorSd <= 0
      ? tumorPoint >= tumorCellsMin
        ? 8
        : -8
      : (tumorPoint - tumorCellsMin) / tumorSd;
  const pPurity = 0.5 * (1 + erfApprox(zPurity / Math.SQRT2));
  const pCells = 0.5 * (1 + erfApprox(zCells / Math.SQRT2));
  return Math.max(0, Math.min(1, pPurity * pCells));
}

function rectIou(
  a: readonly [number, number, number, number],
  b: readonly [number, number, number, number],
): number {
  const ix0 = Math.max(a[0], b[0]);
  const iy0 = Math.max(a[1], b[1]);
  const ix1 = Math.min(a[2], b[2]);
  const iy1 = Math.min(a[3], b[3]);
  if (ix0 >= ix1 || iy0 >= iy1) return 0;
  const inter = (ix1 - ix0) * (iy1 - iy0);
  const areaA = Math.max((a[2] - a[0]) * (a[3] - a[1]), 0);
  const areaB = Math.max((b[2] - b[0]) * (b[3] - b[1]), 0);
  const union = areaA + areaB - inter;
  return inter / Math.max(union, 1e-9);
}

export function suggestCandidates(
  tiles: TileArrays,
  threshold: ThresholdProfile,
  opts: { windowTiles?: number; topK?: number; nmsIou?: number } = {},
): ClientCandidate[] {
  const grid = tiles.grid;
  const nx = grid.grid_nx;
  const ny = grid.grid_ny;
  const wt = Math.min(opts.windowTiles ?? 5, nx, ny);
  if (wt < 1) return [];
  const topK = opts.topK ?? 5;
  const nmsIou = opts.nmsIou ?? 0.3;

  const sanitize = (a: Float32Array) => {
    const out = new Float32Array(a.length);
    for (let i = 0; i < a.length; i++) {
      out[i] = Number.isFinite(a[i]) ? a[i] : 0;
    }
    return out;
  };
  const p = sanitize(tiles.purity);
  const sp = sanitize(tiles.purity_sd);
  const n = sanitize(tiles.nuclei);
  const sn = sanitize(tiles.nuclei_sd);
  const tf = sanitize(tiles.tissue_fraction);

  const totalGrid = new Float32Array(nx * ny);
  const tumorGrid = new Float32Array(nx * ny);
  const snVarGrid = new Float32Array(nx * ny);
  const spVarGrid = new Float32Array(nx * ny);
  for (let i = 0; i < nx * ny; i++) {
    const eff = n[i] * tf[i];
    totalGrid[i] = eff;
    tumorGrid[i] = eff * p[i];
    snVarGrid[i] = (sn[i] * tf[i]) ** 2;
    spVarGrid[i] = (sp[i] * p[i]) ** 2;
  }

  // Build integral images for fast window sums.
  function integral(src: Float32Array): Float64Array {
    const out = new Float64Array((nx + 1) * (ny + 1));
    for (let y = 1; y <= ny; y++) {
      let rowSum = 0;
      for (let x = 1; x <= nx; x++) {
        rowSum += src[(y - 1) * nx + (x - 1)];
        out[y * (nx + 1) + x] =
          rowSum + out[(y - 1) * (nx + 1) + x];
      }
    }
    return out;
  }
  const I_total = integral(totalGrid);
  const I_tumor = integral(tumorGrid);
  const I_snVar = integral(snVarGrid);
  const I_spVar = integral(spVarGrid);

  function windowSum(I: Float64Array, ix: number, iy: number): number {
    const x0 = ix;
    const y0 = iy;
    const x1 = ix + wt;
    const y1 = iy + wt;
    const s =
      I[y1 * (nx + 1) + x1] -
      I[y0 * (nx + 1) + x1] -
      I[y1 * (nx + 1) + x0] +
      I[y0 * (nx + 1) + x0];
    return s;
  }

  type Raw = ClientCandidate & { rawScore: number };
  const raw: Raw[] = [];
  for (let iy = 0; iy + wt <= ny; iy++) {
    for (let ix = 0; ix + wt <= nx; ix++) {
      const total = windowSum(I_total, ix, iy);
      if (total <= 0) continue;
      const tumor = windowSum(I_tumor, ix, iy);
      const tumorSd = Math.sqrt(
        Math.max(windowSum(I_snVar, ix, iy) + windowSum(I_spVar, ix, iy), 0),
      );
      const puritySd = Math.sqrt(
        Math.max(windowSum(I_spVar, ix, iy) / Math.max(total * total, 1e-9), 0),
      );
      const purity = tumor / Math.max(total, 1e-9);
      const prob = proxyAdequacyProb(
        purity,
        puritySd,
        tumor,
        tumorSd,
        threshold.purity_min,
        threshold.tumor_cells_min,
      );
      const score = prob + 1e-6 * tumor;
      const x0 = grid.offset_x + ix * grid.stride_x;
      const y0 = grid.offset_y + iy * grid.stride_y;
      const x1 = x0 + wt * grid.stride_x;
      const y1 = y0 + wt * grid.stride_y;
      raw.push({
        rank: 0,
        score,
        rawScore: score,
        bbox_thumb_px: [x0, y0, x1, y1] as const,
        polygon: [
          [x0, y0],
          [x1, y0],
          [x1, y1],
          [x0, y1],
        ],
        purity_point: purity,
        total_nuclei_point: total,
        tumor_nuclei_point: tumor,
        adequacy_probability: prob,
      });
    }
  }
  raw.sort((a, b) => b.score - a.score);
  const selected: ClientCandidate[] = [];
  for (const cand of raw) {
    if (selected.every((s) => rectIou(cand.bbox_thumb_px, s.bbox_thumb_px) < nmsIou)) {
      selected.push({ ...cand, rank: selected.length + 1 });
    }
    if (selected.length >= topK) break;
  }
  return selected;
}

File: enso-cell-count/backend/src/enso_purity/macrodissection/candidates.py
"""Candidate macrodissection ROI auto-suggestion.

The frontend exposes a "candidate areas" panel that lists, for each slide,
the top few rectangular regions that are most promising for macrodissection.
Pathologists can click a candidate to load it as an editable polygon and
then refine the boundary.

The candidate scoring is intentionally simple to keep behaviour predictable:

* Build a sliding window of size ``window_tiles × window_tiles`` over the
  tile grid.
* For each window with non-zero tissue, compute the deterministic point
  estimates (``purity``, ``total_nuclei``, ``tumor_nuclei``).
* Compute a quick adequacy probability proxy from the per-tile σ values
  using a closed-form Gaussian approximation (Monte-Carlo would be too slow
  to score thousands of windows). The result is later confirmed by the full
  MC engine when the user clicks the candidate.
* Rank windows by ``adequacy_probability``, breaking ties on
  ``tumor_nuclei`` (more material is better).
* Apply rectangle non-max-suppression so the top K do not all hug the same
  hotspot.

The function always returns *axis-aligned rectangles* as polygons in
thumbnail-pixel coordinates; the frontend reshapes them into editable
polygons.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

from .roi import Polygon, TileArrays


@dataclass
class CandidateROI:
    rank: int
    score: float
    bbox_thumb_px: tuple[float, float, float, float]  # (x0, y0, x1, y1)
    polygon: list[tuple[float, float]]
    purity_point: float
    total_nuclei_point: float
    tumor_nuclei_point: float
    adequacy_probability: float


def _proxy_adequacy_probability(
    purity_point: float,
    purity_sd: float,
    tumor_nuclei_point: float,
    tumor_nuclei_sd: float,
    *,
    purity_min: float,
    tumor_cells_min: float,
) -> float:
    """Closed-form Gaussian-tail approximation of the adequacy probability.

    Assumes independent purity and tumor-nuclei estimates and a Gaussian noise
    model. The result is conservative when the two are positively correlated
    (which they typically are), so callers can use it as a lower bound.
    """

    if purity_sd <= 0:
        p_purity = 1.0 if purity_point >= purity_min else 0.0
    else:
        z_purity = (purity_point - purity_min) / max(purity_sd, 1e-6)
        p_purity = 0.5 * (1.0 + math.erf(z_purity / math.sqrt(2)))

    if tumor_nuclei_sd <= 0:
        p_cells = 1.0 if tumor_nuclei_point >= tumor_cells_min else 0.0
    else:
        z_cells = (tumor_nuclei_point - tumor_cells_min) / max(tumor_nuclei_sd, 1e-6)
        p_cells = 0.5 * (1.0 + math.erf(z_cells / math.sqrt(2)))

    return float(max(0.0, min(1.0, p_purity * p_cells)))


def _rect_iou(
    a: tuple[float, float, float, float],
    b: tuple[float, float, float, float],
) -> float:
    ax0, ay0, ax1, ay1 = a
    bx0, by0, bx1, by1 = b
    ix0 = max(ax0, bx0)
    iy0 = max(ay0, by0)
    ix1 = min(ax1, bx1)
    iy1 = min(ay1, by1)
    if ix0 >= ix1 or iy0 >= iy1:
        return 0.0
    inter = (ix1 - ix0) * (iy1 - iy0)
    area_a = max((ax1 - ax0) * (ay1 - ay0), 0.0)
    area_b = max((bx1 - bx0) * (by1 - by0), 0.0)
    union = area_a + area_b - inter
    return inter / max(union, 1e-9)


def suggest_candidates(
    tiles: TileArrays,
    *,
    purity_min: float,
    tumor_cells_min: float,
    window_tiles: int = 5,
    top_k: int = 5,
    nms_iou: float = 0.30,
) -> list[CandidateROI]:
    """Return up to ``top_k`` candidate rectangular ROIs."""

    grid = tiles.grid
    nx, ny = grid.grid_nx, grid.grid_ny
    if nx < 1 or ny < 1:
        return []
    wt = min(window_tiles, nx, ny)
    if wt < 1:
        return []

    # Pre-fill NaN-safe arrays.
    p = np.where(np.isnan(tiles.purity), 0.0, tiles.purity).astype(np.float64)
    sp = np.where(np.isnan(tiles.purity_sd), 0.0, tiles.purity_sd).astype(np.float64)
    n = np.where(np.isnan(tiles.nuclei), 0.0, tiles.nuclei).astype(np.float64)
    sn = np.where(np.isnan(tiles.nuclei_sd), 0.0, tiles.nuclei_sd).astype(np.float64)
    tf = np.where(np.isnan(tiles.tissue_fraction), 0.0, tiles.tissue_fraction).astype(
        np.float64
    )

    eff = n * tf
    eff_purity = eff * p

    # Cumulative sum trick for fast window sums.
    def _window_sum(arr: np.ndarray) -> np.ndarray:
        # 2D cumulative sum with zero-padding on top/left.
        cs = np.pad(arr, ((1, 0), (1, 0)), mode="constant").cumsum(axis=0).cumsum(axis=1)
        # Window sums over wt × wt blocks at the top-left corner (iy, ix).
        out_h = ny - wt + 1
        out_w = nx - wt + 1
        if out_h <= 0 or out_w <= 0:
            return np.zeros((0, 0), dtype=np.float64)
        a = cs[wt : wt + out_h, wt : wt + out_w]
        b = cs[:out_h, wt : wt + out_w]
        c = cs[wt : wt + out_h, :out_w]
        d = cs[:out_h, :out_w]
        return a - b - c + d

    total_w = _window_sum(eff)
    tumor_w = _window_sum(eff_purity)
    # For σ propagation we sum variances (independent tile assumption).
    sn_var = (sn * tf) ** 2  # variance of nuclei per tile
    sp_var = (sp * p) ** 2  # rough propagation
    sn_w = _window_sum(sn_var)
    sp_w = _window_sum(sp_var)
    if total_w.size == 0:
        return []

    # Vectorized scoring per window.
    flat_idx = np.arange(total_w.size)
    iy_idx = flat_idx // total_w.shape[1]
    ix_idx = flat_idx % total_w.shape[1]
    candidates_raw: list[CandidateROI] = []
    for k in range(total_w.size):
        iy = int(iy_idx[k])
        ix = int(ix_idx[k])
        total = float(total_w[iy, ix])
        tumor = float(tumor_w[iy, ix])
        if total <= 0:
            continue
        purity_point = tumor / max(total, 1e-9)
        tumor_sd = math.sqrt(max(float(sn_w[iy, ix]) + float(sp_w[iy, ix]), 0.0))
        purity_sd = math.sqrt(max(float(sp_w[iy, ix]), 0.0) / max(total**2, 1e-9))
        prob = _proxy_adequacy_probability(
            purity_point=purity_point,
            purity_sd=purity_sd,
            tumor_nuclei_point=tumor,
            tumor_nuclei_sd=tumor_sd,
            purity_min=purity_min,
            tumor_cells_min=tumor_cells_min,
        )
        # Composite score: probability is primary, tumor nuclei breaks ties.
        score = prob + 1e-6 * tumor
        x0 = grid.offset_x + ix * grid.stride_x
        y0 = grid.offset_y + iy * grid.stride_y
        x1 = x0 + wt * grid.stride_x
        y1 = y0 + wt * grid.stride_y
        candidates_raw.append(
            CandidateROI(
                rank=0,
                score=score,
                bbox_thumb_px=(x0, y0, x1, y1),
                polygon=[(x0, y0), (x1, y0), (x1, y1), (x0, y1)],
                purity_point=purity_point,
                total_nuclei_point=total,
                tumor_nuclei_point=tumor,
                adequacy_probability=prob,
            )
        )

    candidates_raw.sort(key=lambda c: c.score, reverse=True)
    selected: list[CandidateROI] = []
    for cand in candidates_raw:
        if all(_rect_iou(cand.bbox_thumb_px, s.bbox_thumb_px) < nms_iou for s in selected):
            cand.rank = len(selected) + 1
            selected.append(cand)
        if len(selected) >= top_k:
            break
    return selected


def candidate_polygon_to_polygon(cand: CandidateROI) -> Polygon:
    return cand.polygon
